assert/assertions.go

```diff
@@ -521,30 +527,36 @@ func NotSame(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}
 	if h, ok := t.(tHelper); ok {
 		h.Helper()
 	}
-
-	if samePointers(expected, actual) {
-		return Fail(t, fmt.Sprintf(
-			"Expected and actual point to the same object: %p %#v",
-			expected, expected), msgAndArgs...)
+	if same, ok := samePointers(expected, actual); ok {
+		//if ok is true, then both arguments are pointers
+		if same {
+			return Fail(t, fmt.Sprintf(
+				"Expected and actual point to the same object: %p %#v",
+				expected, expected), msgAndArgs...) //because Fail return false, but we need to
+		}
+		return true
 	}
-	return true
+
+	//fails when the arguments are not pointers
+	return !(Fail(t, "Both arguments must be pointers", msgAndArgs...))
+	//we return !Fail because Fail return false, but we need to return true if the arguments are not pointers(by default)
 }
 
 // samePointers compares two generic interface objects and returns whether
 // they point to the same object
-func samePointers(first, second interface{}) bool {
+func samePointers(first, second interface{}) (same bool, ok bool) {
```

Comment:
The doc-string for the internal samePointers needs updating. It wasn't strictly correct before as it doesn't mention the type checking.